src/include/storage/block_compactor.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include "storage/arrow_block_metadata.h"
+#include "storage/data_table.h"
+#include "storage/storage_defs.h"
+#include "transaction/transaction_manager.h"
+namespace terrier::storage {
+
+/**
+ * Typedef for a standard hash map with varlen entry as the key. The map uses deep equality checks (whether
+ * the stored underlying byte string is the same) for key comparison.
+ */
+template <class T>
+using VarlenEntryMap = std::unordered_map<VarlenEntry, T, VarlenContentHasher, VarlenContentDeepEqual>;
+
+/**
+ * The block compactor is responsible for taking hot data blocks that are considered to be cold, and make them
+ * arrow-compatible. In the process, any gaps resulting from deletes or aborted transactions are also eliminated.
+ * If the compaction is successful, the block is considered to be fully cold and will be accessed mostly as read-only
+ * data.
+ */
+class BlockCompactor {
+ private:
+  // A Compaction group is a series of blocks all belonging to the same data table. We compact them together
+  // so slots can be freed up. If we only eliminate gaps, deleted slots will never be reclaimed.
```

Comment:
Can you clarify the distinction between "compacting them" and "eliminat[ing] gaps"? Compacting means across blocks while eliminating gaps is just within blocks in this terminology?